### DISCUSSION TITLE
Is it possible to fit a variogram on 2D image and generate an equivalent 3D field using gstools?

### DISCUSSION THREAD

**Discussion by rkenko** [GeoStat-Framework/GSTools#287] [Q&A] [answered]
Hi, 

I am new in geostats and i wonder if it is possible to fit a variogram on a 2D field then generate a 3D equivalent field.

This is the code i am using: 

```python
#importing the image from which we seek the variogram 
image       = plt.imread("field.png")[:,:,0]
# spreading values of imported image between ]-inf;+inf[
image       = np.tan(((10/np.pi)*(image))+(np.pi/2)) 


# estimating empirical variogram
gamma = gs.vario_estimate_structured(image)
bin_center = np.arange(len(gamma))


# fiiting the empirical variogram at small range, 128 px
bins        = np.arange(128)
fit_model   = gs.Gaussian(dim=2)
fit_model.fit_variogram(bin_center[:128], gamma[:128], nugget=False)
# printing all the parameters of the covariance model fitted
print(fit_model)


# generating 3D field from fitted variogram with size 256
x  = y  = z = range(256)
model       = gs.Gaussian(dim=3, var=fit_model.var, len_scale=fit_model.len_scale)
# generating the spatial random field
srf         = gs.SRF(model)
srf((x, y, z), mesh_type = 'structured')
srf         = srf.field


# passing the field through atan in order to spread distribution between [0;1]
srf         = (np.pi/10)*(np.arctan(srf)-(np.pi/2))+1
print(np.amin(srf))
print(np.amax(srf))
plt.imsave("x_middle_slice.png",srf[:,:,128],cmap="gray",vmin=0,vmax=1)
# saving to a VTK file for visualization
srf         = pv.wrap(srf)
srf.save('gs.vtk')
```

My question is: am i doing right ? the purpose of the tanh is to ensure values stays in the range [0,1] since they represents pixels values in a picture. without the tanh, the field isn't realistic in terms of colors.

I observed the slices of the generated 3D volume dont have the same variogram than the 2D picture of the begining. I expected them to be the same... Is there a way to correct my code please ?

**Comment by MuellerSeb** [comment, marked as answer]
Hi there,

This looks right to me. I would check for normality of the data after the tan transformation since we need a normal distribution to apply this workflow.

You could adopt two things:
1. I would use `vario_estimate` and not `vario_estimate_structured` to get an omnidirectional estimation for the empirical variogram:
   ```python
   x = y = np.arange(image.shape[0]), np.arange(image.shape[1])
   bin_center, gamma = gs.vario_estimate((x,y), image, mesh_type="structured")
   ```
2. You can use the 3D model directly to fit the omnidirectional variogram
   ```python
   model = gs.Gaussian(dim=3)
   model.fit_variogram(bin_center, gamma, nugget=False)
   ```
The rest looks good to me :wink:

Cheers,
Sebastian

**Reply by rkenko** [reply]
Hi @MuellerSeb , thank you for your time. I tried to implement your suggestion.

I have a technical question and 2 theorical others. This is my code so far:

```python
#importing the image from which we seek the variogram 
image       = plt.imread("field.png")[:,:,0]
# spreading values of imported image between ]-inf;+inf[
image       = np.tan(((10/np.pi)*(image))+(np.pi/2)) 


# estimating empirical variogram
bins = np.arange(128)
x = y = image.shape[0], image.shape[1]
bin_center, gamma = gs.vario_estimate((x, y), image,bins)


# fiiting the empirical variogram at small range, 128 px
fit_model   = gs.Gaussian(dim=3)
fit_model.fit_variogram(bin_center[:128], gamma[:128], nugget=False)
# printing all the parameters of the covariance model fitted
print(fit_model)


# generating 3D field from fitted variogram with size 256
x  = y  = z = range(256)
model       = gs.Gaussian(dim=3, var=fit_model.var, len_scale=fit_model.len_scale)
# generating the spatial random field
srf         = gs.SRF(model)
srf((x, y, z), mesh_type = 'structured')
srf         = srf.field


# passing the field through atan in order to spread distribution between [0;1]
srf         = (np.pi/10)*(np.arctan(srf)-(np.pi/2))+1
plt.imsave("x_middle_slice.png",srf[:,:,128],cmap="gray",vmin=0,vmax=1)
# saving to a VTK file for visualization
srf         = pv.wrap(srf)
srf.save('gs.vtk')
```

My technical question question is: using this code i have this error "ValueError: Formatting: position tuple doesn't match dimension." > What is the good way to use vario_estimate on image dataset with gstools ? 

My theorical question is: what are de conditions on the field before the computation of the variogram for my task ? (generation of a 3d field after the fit). when you say "check the normality", you mean that my field need to be 0_centered with unit variance ? 

You you have an implementation to fit the empirical variogram of a non-stationary field ? I ideally want to fit at small range and to end up with something like in the figure in order to generate a better 3d field....

source for the image: https://www.researchgate.net/publication/233635506_Teacher%27s_Aide_Modeling_Hole-Effect_Variograms_of_Lithology-Indicator_Variables/figures?lo=1

![non_stationary_fit](https://user-images.githubusercontent.com/107155677/206743266-6d168cce-da2f-489e-a40c-1b2d981a9974.png)

**Reply by MuellerSeb** [reply]
For the technical question: the `mesh_type` is crucial:
```python
bin_center, gamma = gs.vario_estimate((x,y), image, mesh_type="structured")
```

1. theoretical question:
Just look at the histogram. It should just look rather normal (not skewed or something). It doesn't need to be standardized.

2. theoretical question: Have a look at the JBessel covariance model. That is a holemodel and should be what you are looking for: https://geostat-framework.readthedocs.io/projects/gstools/en/v1.4.0/generated/gstools.covmodel.JBessel.html

Cheers, Sebastian

**Reply by rkenko** [reply]
Hi @MuellerSeb, thank you for yout time. 

I have one last remark. The code is not working using:

```python
x = y = image.shape[0], image.shape[1]
bin_center, gamma = gs.vario_estimate((x,y), image, mesh_type="structured")
```
Here is the error:
```
Traceback (most recent call last):
  File "/home/generate_grf.py", line 96, in <module>
    bin_center, gamma = gs.vario_estimate((x, y), image, mesh_type="structured")
  File "/home/anaconda3/lib/python3.9/site-packages/gstools/variogram/variogram.py", line 270, in vario_estimate
    pos, __, dim = format_struct_pos_shape(
  File "/home/anaconda3/lib/python3.9/site-packages/gstools/tools/geometric.py", line 513, in format_struct_pos_shape
    raise ValueError("Formatting: position tuple doesn't match dimension.")
ValueError: Formatting: position tuple doesn't match dimension.

```

The code works using this line. But doing this i wonder if am only giving the values of the diagonal pixels.

```python
x , y = np.arange(image.shape[0]), np.arange(image.shape[1])
bin_center, gamma = gs.vario_estimate((x, y), image,bins)
```

I want to do something like this, but it don't work...
```python
nb_points = 10000
maxy, maxx  = image.shape
x, y = np.random.randint(0, maxx, nb_points), np.random.randint(0, maxy, nb_points)
bin_center, gamma = gs.vario_estimate((x,y), image, mesh_type="structured")
```

**Reply by MuellerSeb** [reply]
Sorry, there was a mistake on my side. You correctly put `np.arange` around the image shapes. But you still need to put `mesh_type="structured"` in the call signature. You don't need `bins` (but you can give `bin_edges` of course), since they are calculated automatically.
If you want to have a sampled variogram (if I interpret your `np.random.randint(0, maxx, nb_points)` correctly), you can do that directly in the estimation routine. So the call could look like this:
```python
x , y = np.arange(image.shape[0]), np.arange(image.shape[1])
bin_center, gamma = gs.vario_estimate((x, y), image, mesh_type="structured", sampling_size=10000)
```

**Reply by rkenko** [reply]
yes, i copy-pasted the wrong code, i used mesh_type="structured" for this test. Thanks a lot

**Reply by MuellerSeb** [reply]
One side-note: take a look at the image, since I think the image is not given in `xy` coordinates but in matrix style `ij`. So it could be, you need to transpose the image after loading it
```python
image = image.T
```
